src/Archives/ArchiveUnpacker.h

```diff
@@ -20,15 +20,16 @@ namespace Archives
 		uint64_t GetVolumeFileSize() { return m_ArchiveFileSize; };
 		int GetNumberOfPackedFiles() { return m_NumberOfPackedFiles; };
 		bool ContainsFile(const std::string& filename);
+		// Returns -1 if internalFilename is not present in archive.
+		int GetInternalFileIndex(const std::string& internalFilename);
+		void ExtractFile(const std::string& internalFilename, const std::string& pathOut);
```

Comment:
This is one of those base classes where I had a mix of virtual methods with non-virtual methods, and a handful of member variables. I kind of regretted some aspects of the design later. As a note for going forward, my design of this base class might not be such a great example to base future code on.

In terms of the current changes, you seem to have done a pretty decent job with the breakdown of what should be virtual and what should be non-virtual. As much as the original implementation still worries me, I think I actually agree with what you've done here.

Reply:
I think virtual classes are a fine solution in this case. It was more the multiple inheritance of 2 virtual classes into ArchiveFile I was/am unsure about.